File: Java/li1/plp/imperative1/Exemplo1.java
package li1.plp.imperative1;

import li1.plp.expressions2.expression.Id;
import li1.plp.expressions2.expression.ValorInteiro;
import li1.plp.imperative1.command.Comando;
import li1.plp.imperative1.command.ComandoDeclaracao;
import li1.plp.imperative1.command.Write;
import li1.plp.imperative1.declaration.DeclaracaoVariavel;
import li1.plp.imperative1.memory.ContextoCompilacaoImperativa;
import li1.plp.imperative1.memory.ContextoExecucaoImperativa;
import li1.plp.imperative1.memory.ListaValor;
import li1.plp.expressions2.memory.IdentificadorJaDeclaradoException;
import li1.plp.expressions2.memory.IdentificadorNaoDeclaradoException;
import li1.plp.imperative1.memory.EntradaVaziaException;
import li1.plp.imperative1.memory.ErroTipoEntradaException;


public class Exemplo1 {

    public static void main(String[] args) {
        Id idA = new Id("a");

        // var a = 3
        DeclaracaoVariavel decVarA = new DeclaracaoVariavel(idA, new ValorInteiro(3));

        // write(a)
        Comando writeA = new Write(idA);

        // { var a = 3; write(a) }
        Comando blocoDeclaracao = new ComandoDeclaracao(decVarA, writeA);

        Programa programa = new Programa(blocoDeclaracao);

        try {
            ContextoCompilacaoImperativa ambComp = new ContextoCompilacaoImperativa(new ListaValor());
            if (programa.checaTipo(ambComp)) {
                ContextoExecucaoImperativa ambExec = new ContextoExecucaoImperativa(new ListaValor());
                System.out.println("Saida do Exemplo 1: " + programa.executar(ambExec));
            } else {
                System.out.println("Erro de tipo no Exemplo 1!");
            }
        } catch (IdentificadorJaDeclaradoException e) {
            System.err.println("Erro!! Identificador já declarado!");
            System.err.println("Detalhes: " + e.getMessage());
            
        } catch (IdentificadorNaoDeclaradoException e) {
            System.err.println("Erro!! Identificador não declarada!");
            System.err.println("Detalhes: " + e.getMessage());
            
        } catch (EntradaVaziaException e) {
            System.err.println("Erro! Nenhuma entrada foi fornecida!");
            System.err.println("Detalhes: " + e.getMessage());
            
        } catch (ErroTipoEntradaException e) {
            System.err.println("Erro! O tipo de dado na entrada nao corresponde!");
            System.err.println("Detalhes: " + e.getMessage());
        }

    }
}

File: Python/li1-python/Exemplo1.py
from expression.id import Id
from expression.valor_inteiro import ValorInteiro
from command.comando_declaracao import ComandoDeclaracao
from command.write import Write
from declaration.declaracao_variavel import DeclaracaoVariavel
from memory.contexto_compilacao_imperativa import ContextoCompilacaoImperativa
from memory.contexto_execucao_imperativa import ContextoExecucaoImperativa
from memory.lista_valor import ListaValor
from Programa import Programa
from memory.identificador_ja_declarado_exception import IdentificadorJaDeclaradoException
from memory.identificador_nao_declarado_exception import IdentificadorNaoDeclaradoException
from memory.entrada_vazia_exception import EntradaVaziaException
from memory.erro_tipo_entrada_exception import ErroTipoEntradaException

def main():
    idA = Id("a")

    # var a = 3
    decVarA = DeclaracaoVariavel(idA, ValorInteiro(3))

    # write(a)
    writeA = Write(idA)

    # { var a = 3; write(a) }
    blocoDeclaracao = ComandoDeclaracao(decVarA, writeA)

    programa = Programa(blocoDeclaracao)

    try:
        ambComp = ContextoCompilacaoImperativa(ListaValor())
        if programa.checaTipo(ambComp):
            ambExec = ContextoExecucaoImperativa(ListaValor())
            print("Saida do Exemplo 1:", programa.executar(ambExec))
        else:
            print("Erro de tipo no Exemplo 1!")
            
    except IdentificadorJaDeclaradoException as e:
        print("Erro!! Identificador já declarado!")
        print("Detalhes:", e)
    except IdentificadorNaoDeclaradoException as e:
        print("Erro!! Identificador não declarada!")
        print("Detalhes:", e)
    except EntradaVaziaException as e:
        print("Erro! Nenhuma entrada foi fornecida!")
        print("Detalhes:", e)
    except ErroTipoEntradaException as e:
        print("Erro! O tipo de dado na entrada nao corresponde!")
        print("Detalhes:", e)

if __name__ == '__main__':
    main()
